### PR TITLE
Add shim CLI and sequencer

### LINKED ISSUE #101: Create an API and  a CLI for realtime_zshimming sequencer and optimizer
## What is realtime_zhim
TODO

## API
The API will:
- resample coil profile into the space of anat
- resample (in time) the physio trace to the 4d fieldmap data so that each time point of the fieldmap has its corresponding respiratory probe value.
- run optimizer within mask.nii
- outputs:
  - fieldmap_shimmed.nii
  - coefficients.csv
  - figures

Inputs
- Fieldmap
- Anat
- PMU
- Mask
- Output
- Coil Profile

## CLI

CLI name suggestion
- st_realtime_zshim

CLI flags
- fmap
- anat
- resp
- mask
- output
- coil_profile

Motivated by #37

### PR DESCRIPTION
**Pull request by po09i**
## Description
Adds the ability to shim using the sequencer through `st_b0shim`. More precisely, the PR adds:

- `st_b0shim` CLI
- `st_b0shim` argument to use the sequencer (`st_b0shim static`, `st_b0shim realtime`, `st_b0shim gradient_realtime`)
- `st_b0shim` argument to define slices (`st_b0shim define_slices`) (Removed)
- Ability to define the slices using the "--slices" and "--slice-factor" options
- Define coil names in the constraints when defining a `Coil`
- Displays the R2 score of the linear fit in debug mode for rt sequencer
- Ability to output shim coefficients in different file formats
  - For realtime, available file formats: `slicewise-ch`, `chronological-ch`, `eva`
  - This allows for custom coils and scanner coils.
- Function to convert from physical CS to Gradient CS

## To test
Run the tests in test/cli/test_cli_b0shim

## Related
Fixes #101 
Related to https://github.com/shimming-toolbox/shimming-toolbox/issues/297

### CODE REVIEW

**Review by gaspardcereza — APPROVED** [review]
Tested it locally and it works just fine. I went through the code and didn't spot any mistake/typo. LGTM !

File: shimmingtoolbox/coils/siemens_basis.py
Comment (gaspardcereza) [inline shimmingtoolbox/coils/siemens_basis.py:155]: You could add a comment explaining what's wrong with the orders (or refactor `_get_scaling_factors` and `_reorder_to_siemens` as suggested)

File: shimmingtoolbox/coils/siemens_basis.py
Comment (gaspardcereza) [inline shimmingtoolbox/coils/siemens_basis.py:13]: Its a lot of precision for something that changes depending on the surroundings of the protons. Do we need to be that precise ? Also the value might differ slightly depending on the source...

File: config/coil_config.json
Comment (gaspardcereza) [inline config/coil_config.json:3]: Are these values provided by Siemens ? Will the user that want to use a different coil setup have to create a similar file ?

File: shimmingtoolbox/utils.py
Comment (gaspardcereza) [inline shimmingtoolbox/utils.py]: Should it be a string (with '') or just the word alone (i.e. `'info'` or `info`)?

File: test/shim/test_sequencer.py
Comment (gaspardcereza) [inline test/shim/test_sequencer.py:216]: Try to have 100% test coverage before merging (this way we won't have to do it later)

File: shimmingtoolbox/coils/siemens_basis.py
Comment (po09i) [reply shimmingtoolbox/coils/siemens_basis.py:13]: It seems better to add more precision than not having that precision at all even if it might differ slightly

File: config/coil_config.json
Comment (po09i) [reply config/coil_config.json:3]: The value are currently arbitrary. The idea is indeed to have the user fill out this example file when they set up a new coil to reflect the min/max parameters.

File: shimmingtoolbox/utils.py
Comment (po09i) [reply shimmingtoolbox/utils.py]: It should be a string. I updated the doc.

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
             help="Shim according to the specified algorithm as an argument e.g. st_b0shim xxxxx")
```

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
              help="Static B0 fieldmap.")
```

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]: simplify wording as in the comment above

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]: should work if using 2D data

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]: allowed values should be defined in the click arguments

Comment (jcohenadad) [reply shimmingtoolbox/cli/b0shim.py]: and remove choice from usage

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]: fragile when >2 orders will be implemented

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
              help="Number of slices per shimmed group. For example, if the value is '3', then with the 'sequential' mode, shimming will be performed independently on the following groups: {0,1,2}, {3,4,5}, etc. With the mode 'interleaved', it will be: {0,2,4}, {6,8,10}, etc.")
```

Comment (jcohenadad) [reply shimmingtoolbox/cli/b0shim.py]: possible issue: if total number of slice is not divisible by the nb of slice_factor --> issue to open-- more investigations needed

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
              show_default=True, help="Kernel used to dilate the mask to expand the ROI")
```

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]: i would remove this option and just hard-code sphere

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
              help="Number of voxels to consider outside of the masked area. For example, when doing dynamic shimming with a linear gradient, the coefficient corresponding to the gradient orthogonal to a single slice cannot be estimated: there must be at least 2 (ideally 3) points to properly estimate the linear term. When using 2nd order or more, more dilation is necessary.")
```

File: shimmingtoolbox/cli/b0shim.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
              show_default=True, help="Syntax used to describe the sequence of shim events. Use 'slicewise' to output in row 1, 2, 3, etc. the shim coefficients for slice 1, 2, 3, etc. Use 'chronological' to output in row 1, 2, 3, etc. the shim value for trigger 1, 2, 3, etc. The trigger is an event sent by the scanner and captured by the controller of the shim amplifier. Use 'ch' to output one file per coil channel (coil1_ch1.txt, coil1_ch2.txt, etc.). Use 'coil' to output one file per coil system (coil1.txt, coil2.txt). In the latter case, all coil channels are encoded across multiple columns in the text file.")
```

**Comment by po09i** [comment]
- Name of method: throw away fieldmap
- Happy with st_b0shim?

File: shimmingtoolbox/coils/coil.py
Comment (gaspardcereza) [inline shimmingtoolbox/coils/coil.py]: Is this comma necessary ?

File: shimmingtoolbox/masking/mask_utils.py
Comment (gaspardcereza) [inline shimmingtoolbox/masking/mask_utils.py:7]: "import" should come before "from"

File: shimmingtoolbox/shim/realtime_shim.py
Comment (gaspardcereza) [inline shimmingtoolbox/shim/realtime_shim.py]: Why the same multiplicator for mT and uT ? Do you want to keep the second line ?

File: test/shim/test_sequencer.py
Comment (gaspardcereza) [inline test/shim/test_sequencer.py]: library imports should come before individual functions are imported

File: test/shim/test_sequencer.py
Comment (gaspardcereza) [inline test/shim/test_sequencer.py]: Same question as for previous comma

File: test/shim/test_sequencer.py
Comment (gaspardcereza) [inline test/shim/test_sequencer.py:222]: Do you plan to implement this test later ? Otherwise I would just remove it

File: test/test_siemens_basis.py
Comment (gaspardcereza) [inline test/test_siemens_basis.py]: Why the line jump ?

File: shimmingtoolbox/shim/realtime_shim.py
Comment (po09i) [reply shimmingtoolbox/shim/realtime_shim.py]: Uncommenting the second line will output in uT/m, probably not necessary anymore.

File: test/shim/test_sequencer.py
Comment (po09i) [reply test/shim/test_sequencer.py:222]: Yes, this will eventually be added

File: shimmingtoolbox/coils/coil.py
Comment (po09i) [reply shimmingtoolbox/coils/coil.py]: Resolved in 1a5ce6c.

File: shimmingtoolbox/masking/mask_utils.py
Comment (po09i) [reply shimmingtoolbox/masking/mask_utils.py:7]: Resolved in 1a5ce6c.

File: shimmingtoolbox/shim/realtime_shim.py
Comment (po09i) [reply shimmingtoolbox/shim/realtime_shim.py]: Resolved in 1a5ce6c.

File: test/shim/test_sequencer.py
Comment (po09i) [reply test/shim/test_sequencer.py]: Resolved in 1a5ce6c.

File: test/shim/test_sequencer.py
Comment (po09i) [reply test/shim/test_sequencer.py]: Resolved in 1a5ce6c.

File: test/test_siemens_basis.py
Comment (po09i) [reply test/test_siemens_basis.py]: Resolved in 1a5ce6c.

File: shimmingtoolbox/optimizer/lsq_optimizer.py
Comment (po09i) [inline shimmingtoolbox/optimizer/lsq_optimizer.py:179]: When optimizing, scipy raises a warning if the optimization reaches the bounds. This causes 2 problem, one is that this raises warnings for the tests. It also clutters the output on the terminal for what is an "expected" behaviour (That's why we have bounds). I'm wondering if this is acceptable practice to silence this warning. @jcohenadad @joshuacwnewton

Comment (jcohenadad) [reply shimmingtoolbox/optimizer/lsq_optimizer.py:179]: usually warnings are there for a reason, but in this case I understand that it has to do with the optimization process and that it _could_ be ignored as long as the information is always irrelevant for the user.

For example: could we imagine a scenario where an optimization method would reach the bounds for many channels, and that solution would be less optimal than with another optimization method, in which case the user might want to know that they should maybe try using another optimization method?  Alternatively, as part of the output figures, you could output a graphs with all channels and their shim voltage, and showing the bounds on the graphs, so the user would have a quick overview about the number of maxed out channels (and then decide whether they want to try another method)

Comment (joshuacwnewton) [reply shimmingtoolbox/optimizer/lsq_optimizer.py:179]: My apologies, I missed this ping! I don't have prior experience with scipy's optimization tools, so it's tricky for me to comment on what practices are typical. (I also don't have much prior exposure to section of ST's codebase.)

But, if you still need another set of eyes on this, please feel free to re-ping me and I can try to dig into this warning a little more.

**Pull request merged**
